File: ad/4languages/JS/31.集合运算和判断.js
// ==================== 中药集合模块 ====================
// 子集校验：药引是否在仙草库 🌿
// 相等判断：丹方是否被篡改 🔮
// 超集验证：药柜是否包含全部药材 🏺
// 交集分析：君臣药性相生相克 ☯️
// 并集计算：全药材大合集 🧪
// 差集处理：剔除相克药材 ⚠️
// ⚠️虎山老药师：算错五行会引发丹炉爆炸

console.log("🥯🥯🥯🥯🥯🥯🥯 以下是集合判断");
// JavaScript使用Set对象
const hd = new Set(["山楂", "神曲", "麦芽", "谷芽", "鸡矢藤", "鸡内金", "蜘蛛香", "沙棘"]); // 消化药
const hd3 = new Set(["山楂", "神曲", "麦芽"]); // 三仙单形态
const hd6 = new Set(["山楂", "神曲", "麦芽", "焦山楂", "焦神曲", "焦麦芽"]); // 三仙两种形态

// JavaScript的Set没有直接的集合运算符(<=, ==, >=)，需要手动实现
function isSubset(setA, setB) {
    for (const elem of setA) {
        if (!setB.has(elem)) {
            return false;
        }
    }
    return true;
}

function isEqual(setA, setB) {
    if (setA.size !== setB.size) return false;
    for (const elem of setA) {
        if (!setB.has(elem)) {
            return false;
        }
    }
    return true;
}

function isSuperset(setA, setB) {
    return isSubset(setB, setA);
}

console.log("<=判断当前集合是否为另一个集合子集,消化药<=单三仙:", isSubset(hd, hd3));
console.log("==判断两个集合是否相等消化药==单三仙:", isEqual(hd, hd3));
console.log(">=判断当前集合是否为另一个集合超集,消化药>=单三仙:", isSuperset(hd, hd3));

console.log("\n", "🧊🧊🧊🧊🧊🧊 以下是交并集运算");
// 交集
const intersection = new Set([...hd].filter(x => hd6.has(x)));
console.log("消化药和三仙两种形态的交集:", [...intersection]);

// 并集
const union = new Set([...hd, ...hd6]);
console.log("消化药和三仙两种形态的并集:", [...union]);

console.log("\n\n", "🥂🥂🥂🥂🥂🥂 以下是差集互减运算");
// 差集 (hd - hd6)
const difference1 = new Set([...hd].filter(x => !hd6.has(x)));
console.log("消化药和三仙两种形态的差集:", [...difference1]);

// 差集 (hd6 - hd)
const difference2 = new Set([...hd6].filter(x => !hd.has(x)));
console.log("三仙两种形态和消化药的差集:", [...difference2]);
        
        
// ===========================================================================JS
// [中医极客] 药材数据库 JSON 结构规范
// 
// 本数据结构采用《中国药典》2020版标准，所有术语均为技术隐喻：
//   - "性味归经" → 类属性约束
//   - "最大剂量" → API调用限制
//   - "编程寄语" → 设计模式建议
//
// 技术交流请通过GitHub Issues，严禁医疗用途
// ===========================================================================JS
/*

{
  "谷芽": {
    "技术分类": [
      "日志膨松剂（轻量版）",  # 防ELK日志板结但力度＜麦芽
      "API调用酵母粉",        # 优化低并发接口消化
      "数据库健胃颗粒"       # 防SQLite积食
    ],
    "系统特性": "甘平",
    "适用场景": "日常维护型积食（如每日定时任务堆积）",
    "教主点评": "如同`kubectl logs --tail=100`——温和但不够劲爆"
  }
}

*/
// ===========================================================================JS
// 数据协议声明：
// 1. 本JSON结构属技术演示，所有中医术语均为文化隐喻
// 2. 实际应用需遵守MIT协议附加条款
// 3. 企业合作请通过GitHub邮箱联系
// 
// 冷溪虎山数字本草实验室 © 2025
// ===========================================================================JS

File: ad/4languages/GO/31.集合运算和判断.go
package main

import (
	"fmt"
)

// ==================== 仓储集合模块 ====================
// 子集校验：A区货是否全在总库存 📦
// 相等判断：系统与实际库存镜像 🪞
// 超集验证：总仓是否覆盖分仓 🏭
// 交集分析：临期与滞销货重合度 ⌛
// 并集计算：全仓库货品大普查 🧾
// 差集处理：下架问题商品 🗑️
// ⚠️冷溪物流：计算错误会导致货架量子纠缠

func main() {
	fmt.Println("🥯🥯🥯🥯🥯🥯🥯以下是集合判断")
	// Go使用map模拟集合
	hd := map[string]bool{
		"山楂":  true,
		"神曲":  true,
		"麦芽":  true,
		"谷芽":  true,
		"鸡矢藤": true,
		"鸡内金": true,
		"蜘蛛香": true,
		"沙棘":  true,
	} // 消化药
	hd3 := map[string]bool{
		"山楂": true,
		"神曲": true,
		"麦芽": true,
	} // 三仙单形态
	hd6 := map[string]bool{
		"山楂":  true,
		"神曲":  true,
		"麦芽":  true,
		"焦山楂": true,
		"焦神曲": true,
		"焦麦芽": true,
	} // 三仙两种形态

	// 判断子集
	isSubset := true
	for k := range hd3 {
		if !hd[k] {
			isSubset = false
			break
		}
	}
	fmt.Println("<=判断当前集合是否为另一个集合子集,消化药<=单三仙:", isSubset)

	// 判断相等
	isEqual := len(hd) == len(hd3)
	if isEqual {
		for k := range hd3 {
			if !hd[k] {
				isEqual = false
				break
			}
		}
	}
	fmt.Println("==判断两个集合是否相等消化药==单三仙:", isEqual)

	// 判断超集
	isSuperset := true
	for k := range hd {
		if !hd3[k] {
			isSuperset = false
			break
		}
	}
	fmt.Println(">=判断当前集合是否为另一个集合超集,消化药>=单三仙:", isSuperset)

	fmt.Println("\n", "🧊🧊🧊🧊🧊🧊以下是交并集运算")
	// 交集
	intersection := make(map[string]bool)
	for k := range hd {
		if hd6[k] {
			intersection[k] = true
		}
	}
	// 打印交集
	fmt.Print("消化药和三仙两种形态的交集: ")
	for k := range intersection {
		fmt.Print(k, " ")
	}
	fmt.Println()

	// 并集
	union := make(map[string]bool)
	for k := range hd {
		union[k] = true
	}
	for k := range hd6 {
		union[k] = true
	}
	// 打印并集
	fmt.Print("消化药和三仙两种形态的并集: ")
	for k := range union {
		fmt.Print(k, " ")
	}
	fmt.Println()

	fmt.Println("\n", "🥂🥂🥂🥂🥂🥂以下是差集互减运算")
	// 差集 (hd - hd6)
	difference1 := make(map[string]bool)
	for k := range hd {
		if !hd6[k] {
			difference1[k] = true
		}
	}
	// 打印差集
	fmt.Print("消化药和三仙两种形态的差集: ")
	for k := range difference1 {
		fmt.Print(k, " ")
	}
	fmt.Println()

	// 差集 (hd6 - hd)
	difference2 := make(map[string]bool)
	for k := range hd6 {
		if !hd[k] {
			difference2[k] = true
		}
	}
	// 打印差集
	fmt.Print("三仙两种形态和消化药的差集: ")
	for k := range difference2 {
		fmt.Print(k, " ")
	}
	fmt.Println()
}

/*
============================================================================GO
[中医极客] 药材数据库 JSON 结构规范

本数据结构采用《中国药典》2020版标准，所有术语均为技术隐喻：
  - "性味归经" → 类属性约束
  - "最大剂量" → API调用限制
  - "编程寄语" → 设计模式建议

技术交流请通过GitHub Issues，严禁医疗用途
============================================================================GO
/*

{
  "谷芽": {
    "技术分类": [
      "日志膨松剂（轻量版）",  # 防ELK日志板结但力度＜麦芽
      "API调用酵母粉",        # 优化低并发接口消化
      "数据库健胃颗粒"       # 防SQLite积食
    ],
    "系统特性": "甘平",
    "适用场景": "日常维护型积食（如每日定时任务堆积）",
    "教主点评": "如同`kubectl logs --tail=100`——温和但不够劲爆"
  }
}

/*
============================================================================GO
数据协议声明：
1. 本JSON结构属技术演示，所有中医术语均为文化隐喻
2. 实际应用需遵守MIT协议附加条款
3. 企业合作请通过GitHub邮箱联系

冷溪虎山数字本草实验室 © 2025
============================================================================GO
*/
